File: GCD/gcd.m


#import <Foundation/Foundation.h>
#import <string.h>
#import "chrono.h"

size_t sum_for_index(unsigned index)
{
    size_t sum = 0;
    for (unsigned i = 0; i < index; i++)
        sum += i;
    return sum;
}

void compute_sequentially(size_t iterations)
{
    size_t sums[iterations];
    for (int i = 0; i < iterations; i++) {
        sums[i] = sum_for_index(i);
    }
}

void compute_concurrently(size_t iterations)
{
    __block size_t *sums = malloc(iterations * sizeof(size_t));
    dispatch_queue_t queue = dispatch_queue_create("queue", DISPATCH_QUEUE_CONCURRENT);
    dispatch_apply(iterations, queue, ^(size_t index) {
        sums[index] = sum_for_index((unsigned)index);
    });
    free(sums);
}


int main(int argc, const char * argv[])
{

    if (argc != 3)
    {
        NSLog(@"Usage: gcd -[s|p] ITERATIONS");
        return 1;
    }
    
    timeinterval_t begin = now();

    if (strcmp(argv[1], "-s") == 0)
        compute_sequentially(atoi(argv[2]));
    else if (strcmp(argv[1], "-p") == 0)
        compute_concurrently(atoi(argv[2]));
    else   
    {
        NSLog(@"Usage: gcd -[s|p] ITERATIONS");
        return 1;
    }
    
    timeinterval_t end = now();
    
    NSLog(@"time elapsed: %llu ms", time_elapsed(begin, end, MILLISECONDS));
    
    return 0;
}

File: GCD/gcd.swift

import Foundation

guard CommandLine.argc == 3 else {
    print("Usage gcd -[s|p] ITERATIONS")
    exit(1)
}

guard let iterations = Int(CommandLine.arguments[2]) else {
    print("Usage gcd -[s|p] ITERATIONS")
    exit(1)
}

func sumForIndex(index: Int) -> Int {
    var sum = 0
    for i in 0...index {
        sum += i
    }
    return sum
}

var sums: [Int] = []
sums.reserveCapacity(iterations + 1)

let start = DispatchTime.now()

if CommandLine.arguments[1] == "-s" {
    for i in 0...iterations {
        sums.append(sumForIndex(index: i))
    }
} else if CommandLine.arguments[1] == "-p" {
    DispatchQueue.concurrentPerform(iterations: iterations, execute: { (index) in
        sums.append(sumForIndex(index: index))
    })
    
} else {
    print("Usage gcd -[s|p] ITERATIONS")
    exit(1)
}

let end = DispatchTime.now()

print(end.uptimeNanoseconds - start.uptimeNanoseconds) / 1_000_000)


